pystiche/papers/ulyanov_et_al_2016/data.py

```diff
@@ -0,0 +1,273 @@
+
+    texture_images = {
+        "cezanne": DownloadableImage(
+            urljoin(texture_base_ulyanov, "cezanne.jpg"),
+            md5="fab6d360c361c38c331b3ee5ef0078f5",
+        ),
+        "bricks": DownloadableImage(
+            urljoin(base_ulyanov_suppl_texture, "bricks.png"),
+            md5="1e13818e1fbefbd22f110a1c2f781d40",
+        ),
+        "pebble": DownloadableImage(
+            urljoin(base_ulyanov_suppl_texture, "pebble.png"),
+            md5="5b5e5aa6c579e42e268058a94d683a6c",
+        ),
+        "pixelcity_windows2": DownloadableImage(
+            urljoin(base_ulyanov_suppl_texture, "pixelcity_windows2.jpg"),
+            md5="53026a8411e7c26e959e36d3223f3b8f",
+        ),
+        "red-peppers256.o": DownloadableImage(
+            urljoin(base_ulyanov_suppl_texture, "red-peppers256.o.jpg"),
+            md5="16371574a10e0d10b88b807204c4f546",
+        ),
+    }
+    base_johnson = (
+        "https://raw.githubusercontent.com/jcjohnson/fast-neural-style/master/images/"
+    )
+    style_base_johnson = urljoin(base_johnson, "styles/")
+
+    base_ulyanov_suppl_style = "https://raw.githubusercontent.com/DmitryUlyanov/texture_nets/texture_nets_v1/supplementary//stylization_models/"
+    style_images = {
+        "candy": DownloadableImage(
+            urljoin(style_base_johnson, "candy.jpg"),
+            md5="00a0e3aa9775546f98abf6417e3cb478",
+        ),
+        "the_scream": DownloadableImage(
+            urljoin(style_base_johnson, "the_scream.jpg"),
+            md5="619b4f42c84d2b62d3518fb20fa619c2",
+        ),
+        "Jean Metzinger": DownloadableImage(
+            "https://upload.wikimedia.org/wikipedia/commons/c/c9/Robert_Delaunay%2C_1906%2C_Portrait_de_Metzinger%2C_oil_on_canvas%2C_55_x_43_cm%2C_DSC08255.jpg",
+            title="Portrait of Jean Metzinger",
+            author="Jean Metzinger",
+            date="1906",
+            license=PublicDomainLicense(1906),
+            md5="3539d50d2808b8eec5b05f892d8cf1e1",
+            file="jean_metzinger.jpg",
+        ),
+        "mosaic": DownloadableImage(
+            urljoin(base_ulyanov_suppl_style, "mosaic.jpg"),
+            md5="4f05f1e12961cebf41bd372d909342b3",
+        ),
+        "pleades": DownloadableImage(
+            urljoin(base_ulyanov_suppl_style, "pleades.jpg"),
+            md5="6fc41ac30c2852a5454a0ead2f479dc9",
+        ),
+        "starry": DownloadableImage(
+            urljoin(base_ulyanov_suppl_style, "starry.jpg"),
+            md5="c6d94f7962466b2e80a64ae82523242a",
+        ),
+        "turner": DownloadableImage(
+            urljoin(base_ulyanov_suppl_style, "turner.jpg"),
+            md5="7fdd9603a5182dcef23d7fb1c5217888",
+        ),
+    }
+    return DownloadableImageCollection(
+        {**texture_images, **content_images, **style_images},
+        root=root,
+        download=download,
+        overwrite=overwrite,
+    )
+
+
+def ulyanov_et_al_2016_dataset(
```

Comment:
What dataset are we using here? AFAIK no dataset was mentioned in the paper, correct? What dataset is the author using in his implementation?

Reply:
I currently use images from the `Caltech256` dataset to test the program. In Paper and Branch `master`  `ImageNet  ILSVRC2012`  data is used, whereas in the master branch only images with a size `>256` for height and width are used. There is no information whether he used the data from the training, the validation or both. In Branch `texture_nets_v1` only the validation images from `ImageNet  ILSVRC2012`  are used for the training.

Reply:
> I currently use images from the `Caltech256` dataset to test the program.

For your tests this should be OK, but we should try to use the same data as the author in the end. 

I'm a little confused. There are only 4 cases:

- `impl_params==False`, `instance_norm==False`
- `impl_params==False`, `instance_norm==True`
- `impl_params==True`, `instance_norm==False`
- `impl_params==True`, `instance_norm==True`

Could you list all information you can find on the used images for each case. Otherwise its hard to make a decision.

Reply:
`impl_params==True`, `instance_norm==True`:
- images extracted at random from the ` ImageNet ILSVRC2012` data
- Check `image_size` for `height > 256` and `width > 256 ` otherwise take a different image

`impl_params==True`, `instance_norm==False`:
- images extracted from the ` ImageNet ILSVRC2012` validation data
- No random order

`impl_params==False`, `instance_norm==False` || `impl_params==False`, `instance_norm==True`:
- images extracted at random from the ` ImageNet ILSVRC2012` data

Reply:
Thanks for the summary. A few further questions. I'll number the cases with 1 to 4:

1. As mentioned somewhere else, I think the straight forward approach is to simply center crop images. 
2. Same as 1.
3. The author performed random cropping, correct? If yes, does he do it at runtime, i.e. a new random crop for every time the image is processed, or once upfront? 
4. Same as 3.

In all cases I suggest we implement the random selection with a set random seed to make our  implementation reproducible. I can handle this once the rest of this PR is finished.

Reply:
1. Correct, because there is no information about the preprocessing of the images, center crop is now used. 
2. Same as 1.
3. Wrong, in this case the images are resized to the correct size with a `'bilinear'` `scale` and stored in a `hdf5` file for training (once upfront). At the same time the `repr` on `relu4_2` of the images are also stored in this file. 
4. Correct, he uses random cropping during runtime.  So he performs the preprocessing every time a new batch is created.